File: Halachic Times/GeoNames/src/main/java/com/github/geonames/CountryRegion.java
/*
 * Copyright 2012, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.geonames;

import com.vividsolutions.jts.geom.Coordinate;
import com.vividsolutions.jts.geom.Geometry;
import com.vividsolutions.jts.geom.GeometryCollection;

import org.geotools.geojson.geom.GeometryJSON;

import java.awt.Polygon;
import java.awt.geom.Point2D;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import static com.github.geonames.CountryInfo.*;

/**
 * Country region.
 *
 * @author Moshe Waisberg
 */
public class CountryRegion extends Polygon {

    /** Factor to convert coordinate value to a fixed-point integer. */
    public static final double FACTOR_TO_INT = 1e+5;
    /**
     * Factor to convert coordinate value to a fixed-point integer for city
     * limits.
     */
    private static final double CITY_BOUNDARY = 1e+4;

    private final String countryCode;

    /**
     * Constructs a new region.
     */
    public CountryRegion(String countryCode) {
        super();
        if (ISO639_PALESTINE.equals(countryCode)) {
            countryCode = ISO639_ISRAEL;
        }
        this.countryCode = countryCode;
    }

    public String getCountryCode() {
        return countryCode;
    }

    /**
     * Add a location.
     *
     * @param latitude
     *         the latitude.
     * @param longitude
     *         the longitude.
     */
    public void addLocation(double latitude, double longitude) {
        int x = (int) (longitude * FACTOR_TO_INT);
        int y = (int) (latitude * FACTOR_TO_INT);
        addPoint(x, y);
    }

    /**
     * Add a set of locations.
     *
     * @param polygon
     *         the polygon with locations.
     */
    public void addPolygon(Polygon polygon) {
        for (int i = 0; i < polygon.npoints; i++) {
            addPoint(polygon.xpoints[i], polygon.ypoints[i]);
        }
    }

    /**
     * Find the main vertices that represent the border.
     *
     * @param vertexCount
     *         the number of vertices.
     * @return an array of indexes.
     */
    public int[] findMainVertices(final int vertexCount) {
        int[] indexes = new int[vertexCount];
        int r = 0;
        int n = npoints;

        // Find centre.
        long tx = 0;
        long ty = 0;
        for (int i = 0; i < n; i++) {
            tx += xpoints[i];
            ty += ypoints[i];
        }
        double cx = ((double) tx) / n;
        double cy = ((double) ty) / n;

        final double sweepAngle = (2f * Math.PI) / vertexCount;
        double angleStart = -(sweepAngle / 2f);
        double angleEnd;
        double x, y, a, d;
        int farIndex;
        double farDist;

        for (int v = 0; v < vertexCount; v++) {
            angleEnd = angleStart + sweepAngle;
            farDist = Double.MIN_VALUE;
            farIndex = -1;

            for (int i = 0; i < n; i++) {
                x = xpoints[i];
                y = ypoints[i];
                a = Math.atan2(y - cy, x - cx) + Math.PI;
                if ((angleStart <= a) && (a < angleEnd)) {
                    d = Point2D.distanceSq(cx, cy, x, y);
                    if (farDist < d) {
                        farDist = d;
                        farIndex = i;
                    }
                }
            }

            if (farIndex >= 0)
                indexes[r++] = farIndex;

            angleStart += sweepAngle;
        }

        if (r < vertexCount) {
            switch (r) {
                case 0:
                    addPoint((int) (xpoints[0] - CITY_BOUNDARY), (int) (ypoints[0] - CITY_BOUNDARY));
                    addPoint((int) (xpoints[0] + CITY_BOUNDARY), (int) (ypoints[0] + CITY_BOUNDARY));
                case 1:
                    addPoint((int) (xpoints[1] - CITY_BOUNDARY), (int) (ypoints[1] - CITY_BOUNDARY));
                    addPoint((int) (xpoints[1] + CITY_BOUNDARY), (int) (ypoints[1] + CITY_BOUNDARY));
                    return findMainVertices(vertexCount);
            }

            for (int i = r; i < vertexCount; i++)
                indexes[i] = -1;
        }

        return indexes;
    }

    public static CountryRegion toRegion(String countryCode, GeoShape geoShape) throws IOException {
        CountryRegion region = new CountryRegion(countryCode);

        GeometryJSON json = new GeometryJSON();
        Geometry geometry = json.read(geoShape.getGeoJSON());
        geoShape.setGeometry(geometry);

        addGeometry(region, geometry);

        return region;
    }

    private static void addGeometry(CountryRegion region, Geometry geometry) {
        Coordinate[] coordinates = geometry.getCoordinates();
        final int length = coordinates.length;
        Coordinate coordinate;

        for (int i = 0; i < length; i++) {
            coordinate = coordinates[i];
            region.addLocation(coordinate.y, coordinate.x);
        }
    }

    public static List<CountryRegion> toRegions(String countryCode, GeoShape geoShape) throws IOException {
        List<CountryRegion> regions = new ArrayList<>();
        CountryRegion region;

        GeometryJSON json = new GeometryJSON();
        Geometry geometry = json.read(geoShape.getGeoJSON());
        geoShape.setGeometry(geometry);

        if (geometry instanceof com.vividsolutions.jts.geom.Polygon) {
            region = new CountryRegion(countryCode);
            addGeometry(region, geometry);
            regions.add(region);
        } else if (geometry instanceof GeometryCollection) {
            GeometryCollection gc = (GeometryCollection) geometry;
            int count = gc.getNumGeometries();
            for (int i = 0; i < count; i++) {
                region = new CountryRegion(countryCode);
                addGeometry(region, gc.getGeometryN(i));
                regions.add(region);
            }
        }

        return regions;
    }
}

File: Halachic Times/GeoNames/src/main/java/com/github/geonames/CountryRegion.kt
/*
 * Copyright 2012, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.geonames

import com.github.geonames.util.LocaleUtils.ISO639_ISRAEL
import com.github.geonames.util.LocaleUtils.ISO639_PALESTINE
import com.vividsolutions.jts.geom.Geometry
import java.io.IOException
import kotlin.math.round
import org.geotools.geojson.geom.GeometryJSON

/**
 * Country region.
 *
 * @author Moshe Waisberg
 */
class CountryRegion(countryCode: String, geometry: Geometry) {
    val countryCode: String = if (ISO639_PALESTINE == countryCode) {
        ISO639_ISRAEL
    } else {
        countryCode
    }

    val geometries: List<CountryGeometry>
    val maxAreaIndex: Int

    init {
        val num = geometry.numGeometries
        val geometriesCloses = mutableListOf<Geometry>()
        for (n in 0 until num) {
            val geometryN = geometry.getGeometryN(n).close()
            geometriesCloses.add(geometryN)
        }
        val geometries = merge(geometriesCloses).map { CountryGeometry.toFixedInt(it) }
        this.geometries = geometries
        this.maxAreaIndex = findMaxArea(geometries)
    }

    private fun findMaxArea(geometries: List<CountryGeometry>): Int {
        val lastIndex = geometries.lastIndex
        var maxAreaIndex = -1
        var maxArea = 0.0
        for (n in 0..lastIndex) {
            val geometryN = geometries[n]
            val areaN = geometryN.area
            if (areaN > maxArea) {
                maxArea = areaN
                maxAreaIndex = n
            }
        }
        return maxAreaIndex
    }

    /** Merge neighbouring geometries. */
    private fun merge(
        geometries: List<Geometry>,
        distanceNeighbour: Double = DISTANCE_NEIGHBOUR
    ): List<Geometry> {
        if (geometries.size <= 1) return geometries

        val merged = mutableListOf<Geometry?>()
        merged.addAll(geometries.sortedByDescending { it.area })
        val lastIndex = merged.lastIndex

        for (i in 0..lastIndex) {
            var geoI = merged[i] ?: continue
            val envelopeI = geoI.envelope
            for (j in (i + 1)..lastIndex) {
                val geoJ = merged[j] ?: continue
                val envelopeJ = geoJ.envelope
                try {
                    if (envelopeI.distance(envelopeJ) <= distanceNeighbour) {
                        geoI = geoI.join(geoJ)
                        merged[i] = geoI
                        merged[j] = null
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }

        return merged.filterNotNull()
            .map { it.convexHull() }
    }

    companion object {
        /** Factor to convert coordinate value to a fixed-point integer.  */
        internal const val FACTOR_TO_INT = CountryGeometry.FACTOR_TO_INT

        private const val DISTANCE_NEIGHBOUR = 5.0

        @JvmStatic
        @Throws(IOException::class)
        fun toRegion(countryCode: String, geoShape: GeoShape): CountryRegion {
            val json = GeometryJSON()
            val geometry = json.read(geoShape.geoJSON)
            geoShape.geometry = geometry
            return CountryRegion(countryCode, geometry)
        }

        fun toFixedPoint(degrees: Double): Double {
            return round(degrees * FACTOR_TO_INT)
        }

        fun toFixedPointInt(degrees: Double): Int {
            return toFixedPoint(degrees).toInt()
        }
    }
}
